pkg/frontend/querymiddleware/running.go

```diff
@@ -0,0 +1,78 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/services"
+
+	apierror "github.com/grafana/mimir/pkg/api/error"
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+)
+
+func newFrontendRunningMiddleware(frontendState func() services.State, notRunningBackoffDuration time.Duration, maxRetries int, log log.Logger) Middleware {
+	return MiddlewareFunc(func(next Handler) Handler {
+		return frontendRunningMiddleware{
+			frontendState:             frontendState,
+			notRunningBackoffDuration: notRunningBackoffDuration,
+			maxRetries:                maxRetries,
+			log:                       log,
+
+			next: next,
+		}
+	})
+}
+
+type frontendRunningMiddleware struct {
+	frontendState             func() services.State
+	notRunningBackoffDuration time.Duration
+	maxRetries                int
+	log                       log.Logger
+
+	next Handler
+}
+
+func (f frontendRunningMiddleware) Do(ctx context.Context, r Request) (Response, error) {
+	if err := f.waitForRunning(ctx); err != nil {
+		return nil, err
+	}
+
+	return f.next.Do(ctx, r)
+}
+
+func (f frontendRunningMiddleware) waitForRunning(ctx context.Context) (err error) {
+	spanLog, _ := spanlogger.NewWithLogger(ctx, f.log, "waitForRunning")
+	defer spanLog.Finish()
+
+	attempt := 1
+
+	for {
```

Comment:
can we achieve the same as this for loop, but with calling `ctx = context.WithTimeout(f.notRunningBackoffDuration * f.maxRetries)` and using `services.AwaitRunning(ctx)`? With AwaitRunning we don't have to have any logic around what states are valid states and the sleep can be interrupted if the client gives up waiting (not a problem at the moment per-se).

Reply:
Good idea - I've changed this in https://github.com/grafana/mimir/pull/6621/commits/e07d5e9737a4faa0f331860e9462ad4a85c5e9ba. 

It also means we can reframe the configuration option to be "time to wait for frontend to be ready" rather than in terms of backing off and retrying, which feels nicer too.